File: dca-ui/src/utils/tradeConstants.js
/**
 * 交易系统常量服务
 *
 * 从后端获取常量配置，避免前端硬编码
 * 提供缓存机制，减少重复请求
 */

import { getTradeConstants, getExchanges, getSymbols, getModes } from '@/api/dca/tradeConstants'

// 缓存的常量数据
let constantsCache = null
let cacheExpiry = 0
const CACHE_TTL = 5 * 60 * 1000 // 5分钟缓存

// 默认值（作为后备，当API不可用时使用）
const DEFAULT_CONSTANTS = {
  exchanges: {
    allowed: ['BINANCE', 'OKX'],
    binance: 'BINANCE',
    okx: 'OKX'
  },
  symbols: {
    allowed: ['BTCUSDT', 'ETHUSDT', 'SOLUSDT']
  },
  modes: {
    allowed: ['paper', 'shadow', 'live'],
    paper: 'paper',
    shadow: 'shadow',
    live: 'live'
  },
  actions: {
    open: 'OPEN',
    close: 'CLOSE',
    reduce: 'REDUCE',
    hold: 'HOLD'
  },
  positionSides: {
    long: 'long',
    short: 'short',
    net: 'net'
  },
  marginModes: {
    cross: 'cross',
    isolated: 'isolated'
  },
  positionModes: {
    longShort: 'long_short_mode',
    net: 'net_mode'
  },
  orderTypes: {
    market: 'market',
    limit: 'limit'
  },
  orderStatuses: {
    new: 'NEW',
    filled: 'FILLED',
    partiallyFilled: 'PARTIALLY_FILLED',
    canceled: 'CANCELED',
    rejected: 'REJECTED',
    expired: 'EXPIRED'
  }
}

/**
 * 检查缓存是否有效
 */
function isCacheValid() {
  return constantsCache !== null && Date.now() < cacheExpiry
}

/**
 * 获取所有常量配置
 *
 * @param {boolean} forceRefresh - 是否强制刷新缓存
 * @returns {Promise<Object>} 常量配置对象
 */
export async function fetchAllConstants(forceRefresh = false) {
  if (!forceRefresh && isCacheValid()) {
    return constantsCache
  }

  try {
    const response = await getTradeConstants()
    if (response.code === 200 && response.data) {
      constantsCache = response.data
      cacheExpiry = Date.now() + CACHE_TTL
      return constantsCache
    }
  } catch (error) {
    console.warn('Failed to fetch trade constants, using defaults:', error)
  }

  // 返回默认值
  return DEFAULT_CONSTANTS
}

/**
 * 获取支持的交易所列表
 *
 * @returns {Promise<string[]>} 交易所列表
 */
export async function fetchExchanges() {
  const constants = await fetchAllConstants()
  return constants.exchanges?.allowed || DEFAULT_CONSTANTS.exchanges.allowed
}

/**
 * 获取支持的交易对列表
 *
 * @returns {Promise<string[]>} 交易对列表
 */
export async function fetchSymbols() {
  const constants = await fetchAllConstants()
  return constants.symbols?.allowed || DEFAULT_CONSTANTS.symbols.allowed
}

/**
 * 获取支持的运行模式列表
 *
 * @returns {Promise<string[]>} 运行模式列表
 */
export async function fetchModes() {
  const constants = await fetchAllConstants()
  return constants.modes?.allowed || DEFAULT_CONSTANTS.modes.allowed
}

/**
 * 获取特定常量值
 *
 * @param {string} category - 常量类别 (如 'exchanges', 'modes')
 * @param {string} key - 常量键名 (如 'binance', 'paper')
 * @returns {Promise<string>} 常量值
 */
export async function fetchConstant(category, key) {
  const constants = await fetchAllConstants()
  return constants[category]?.[key] || DEFAULT_CONSTANTS[category]?.[key]
}

/**
 * 清除缓存
 */
export function clearConstantsCache() {
  constantsCache = null
  cacheExpiry = 0
}

/**
 * 获取交易所代码（同步方法，使用缓存或默认值）
 *
 * @returns {Object} 交易所代码映射
 */
export function getExchangeCodes() {
  if (constantsCache) {
    return {
      BINANCE: constantsCache.exchanges?.binance || 'BINANCE',
      OKX: constantsCache.exchanges?.okx || 'OKX'
    }
  }
  return {
    BINANCE: DEFAULT_CONSTANTS.exchanges.binance,
    OKX: DEFAULT_CONSTANTS.exchanges.okx
  }
}

/**
 * 获取运行模式代码（同步方法，使用缓存或默认值）
 *
 * @returns {Object} 运行模式代码映射
 */
export function getModeCodes() {
  if (constantsCache) {
    return {
      PAPER: constantsCache.modes?.paper || 'paper',
      SHADOW: constantsCache.modes?.shadow || 'shadow',
      LIVE: constantsCache.modes?.live || 'live'
    }
  }
  return {
    PAPER: DEFAULT_CONSTANTS.modes.paper,
    SHADOW: DEFAULT_CONSTANTS.modes.shadow,
    LIVE: DEFAULT_CONSTANTS.modes.live
  }
}

// 导出默认常量（用于测试或紧急情况）
export { DEFAULT_CONSTANTS }

File: ruoyi-dca/src/main/java/com/ruoyi/dca/constants/TradeConstants.java
package com.ruoyi.dca.constants;

/**
 * 交易系统常量定义
 *
 * 集中管理所有硬编码值，确保前后端一致
 */
public class TradeConstants {

    private TradeConstants() {
        // 私有构造函数，防止实例化
    }

    // ==================== 交易所相关 ====================

    /**
     * 支持的交易所代码
     */
    public static final String EXCHANGE_BINANCE = "BINANCE";
    public static final String EXCHANGE_OKX = "OKX";

    /**
     * V1 版本支持的交易所列表
     */
    public static final java.util.List<String> V1_ALLOWED_EXCHANGES = java.util.List.of(EXCHANGE_BINANCE, EXCHANGE_OKX);

    // ==================== 交易对相关 ====================

    /**
     * V1 版本支持的交易对列表
     */
    public static final java.util.List<String> V1_ALLOWED_SYMBOLS = java.util.List.of("BTCUSDT", "ETHUSDT", "SOLUSDT");

    // ==================== 运行模式相关 ====================

    /**
     * 运行模式：模拟
     */
    public static final String MODE_PAPER = "paper";

    /**
     * 运行模式：影子
     */
    public static final String MODE_SHADOW = "shadow";

    /**
     * 运行模式：实盘
     */
    public static final String MODE_LIVE = "live";

    /**
     * 支持的运行模式列表
     */
    public static final java.util.List<String> ALLOWED_MODES = java.util.List.of(MODE_PAPER, MODE_SHADOW, MODE_LIVE);

    // ==================== 订单相关 ====================

    /**
     * 订单动作：开仓
     */
    public static final String ACTION_OPEN = "OPEN";

    /**
     * 订单动作：平仓
     */
    public static final String ACTION_CLOSE = "CLOSE";

    /**
     * 订单动作：减仓
     */
    public static final String ACTION_REDUCE = "REDUCE";

    /**
     * 订单动作：持有
     */
    public static final String ACTION_HOLD = "HOLD";

    // ==================== 持仓方向相关 ====================

    /**
     * 持仓方向：多仓
     */
    public static final String POSITION_SIDE_LONG = "long";

    /**
     * 持仓方向：空仓
     */
    public static final String POSITION_SIDE_SHORT = "short";

    /**
     * 持仓方向：净持仓（单向持仓模式）
     */
    public static final String POSITION_SIDE_NET = "net";

    // ==================== 保证金模式相关 ====================

    /**
     * 保证金模式：全仓
     */
    public static final String MARGIN_MODE_CROSS = "cross";

    /**
     * 保证金模式：逐仓
     */
    public static final String MARGIN_MODE_ISOLATED = "isolated";

    // ==================== 仓位模式相关 ====================

    /**
     * 仓位模式：双向持仓
     */
    public static final String POSITION_MODE_LONG_SHORT = "long_short_mode";

    /**
     * 仓位模式：单向持仓
     */
    public static final String POSITION_MODE_NET = "net_mode";

    // ==================== 订单类型相关 ====================

    /**
     * 订单类型：市价单
     */
    public static final String ORDER_TYPE_MARKET = "market";

    /**
     * 订单类型：限价单
     */
    public static final String ORDER_TYPE_LIMIT = "limit";

    // ==================== 订单状态相关 ====================

    /**
     * 订单状态：新订单
     */
    public static final String ORDER_STATUS_NEW = "NEW";

    /**
     * 订单状态：已成交
     */
    public static final String ORDER_STATUS_FILLED = "FILLED";

    /**
     * 订单状态：部分成交
     */
    public static final String ORDER_STATUS_PARTIALLY_FILLED = "PARTIALLY_FILLED";

    /**
     * 订单状态：已取消
     */
    public static final String ORDER_STATUS_CANCELED = "CANCELED";

    /**
     * 订单状态：已拒绝
     */
    public static final String ORDER_STATUS_REJECTED = "REJECTED";

    /**
     * 订单状态：已过期
     */
    public static final String ORDER_STATUS_EXPIRED = "EXPIRED";
}
